sklearn/utils/estimator_checks.py

```diff
@@ -1249,6 +1281,12 @@ def check_dtype_object(name, estimator_orig):
     estimator = clone(estimator_orig)
     y = _enforce_estimator_tags_y(estimator, y)
 
+    # ensure n_features of dictionary == n_features of X for SparseCoder
+    if name == "SparseCoder":
```

Comment:
So it seems that the complex things here is to generate a dictionary that depends on `X` ahead of time.

Maybe the least tricky way is to run `check_estimator` not from the common test but from the test module and inherit from `SparseCoder` a fit method that would generate the dictionary at `fit` time.